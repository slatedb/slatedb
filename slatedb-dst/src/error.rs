use slatedb::Error;
use thiserror::Error as ThisError;

#[derive(Debug, ThisError)]
pub(crate) enum DstError {
    #[error("failed to perform an operation on the SQLite state database")]
    SQLiteStateError(#[from] rusqlite::Error),
}

impl From<DstError> for Error {
    fn from(err: DstError) -> Self {
        let msg = err.to_string();
        match err {
            DstError::SQLiteStateError(e) => Error::operation(msg).with_source(Box::new(e)),
        }
    }
}
