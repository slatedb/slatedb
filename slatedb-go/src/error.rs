use slatedb::Error as SlateError;
use std::ffi::{CStr, CString};
use std::os::raw::c_char;

// Error codes that will be exposed to C/Go
#[repr(C)]
#[derive(Debug, Copy, Clone, PartialEq)]
pub enum CSdbError {
    Success = 0,
    InvalidArgument = 1,
    NotFound = 2,
    AlreadyExists = 3,
    IOError = 4,
    InternalError = 5,
    NullPointer = 6,
    InvalidHandle = 7,
}

// Result type for returning both error codes and messages
#[repr(C)]
pub struct CSdbResult {
    pub error: CSdbError,
    pub message: *mut c_char,
}

// Helper functions for error handling
pub fn create_error_result(error: CSdbError, message: &str) -> CSdbResult {
    let c_message =
        CString::new(message).unwrap_or_else(|_| CString::new("Invalid UTF-8").unwrap());
    CSdbResult {
        error,
        message: c_message.into_raw(),
    }
}

pub fn create_success_result() -> CSdbResult {
    CSdbResult {
        error: CSdbError::Success,
        message: std::ptr::null_mut(),
    }
}

pub fn safe_str_from_ptr(ptr: *const c_char) -> Result<&'static str, CSdbError> {
    if ptr.is_null() {
        return Err(CSdbError::NullPointer);
    }

    unsafe {
        CStr::from_ptr(ptr)
            .to_str()
            .map_err(|_| CSdbError::InvalidArgument)
    }
}

pub fn slate_error_to_code(error: &SlateError) -> CSdbError {
    // Use string matching since we can't pattern match on the exact variants
    let error_str = format!("{:?}", error);
    if error_str.contains("NotFound") {
        CSdbError::NotFound
    } else if error_str.contains("AlreadyExists") {
        CSdbError::AlreadyExists
    } else if error_str.contains("InvalidRequest") {
        CSdbError::InvalidArgument
    } else if error_str.contains("IO") {
        CSdbError::IOError
    } else {
        CSdbError::InternalError
    }
}
