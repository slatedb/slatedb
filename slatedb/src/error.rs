use object_store::path::Path;
use std::any::Any;
use std::ops::Bound;
use std::sync::Mutex;
use std::time::Duration;
use std::{path::PathBuf, sync::Arc};
use thiserror::Error as ThisError;
use uuid::Uuid;

use crate::bytes_range::BytesRange;
use crate::merge_operator::MergeOperatorError;

#[non_exhaustive]
#[derive(Clone, Debug, ThisError)]
pub(crate) enum SlateDBError {
    #[error("io error")]
    IoError(#[from] Arc<std::io::Error>),

    #[error("checksum mismatch")]
    ChecksumMismatch,

    #[error("empty SSTable")]
    EmptySSTable,

    #[error("empty block metadata")]
    EmptyBlockMeta,

    #[error("empty block")]
    EmptyBlock,

    #[error("empty manifest")]
    EmptyManifest,

    #[error("object store error")]
    ObjectStoreError(#[from] Arc<object_store::Error>),

    #[error("manifest file already exists")]
    ManifestVersionExists,

    #[error("failed to find manifest with id. id=`{0}`")]
    ManifestMissing(u64),

    #[error("failed to find latest manifest")]
    LatestManifestMissing,

    #[error("invalid deletion")]
    InvalidDeletion,

    #[error("invalid sst error")]
    InvalidFlatbuffer(#[from] flatbuffers::InvalidFlatbuffer),

    #[error("invalid DB state error")]
    InvalidDBState,

    #[error("unsupported operation. operation=`{0}`")]
    Unsupported(String),

    #[error("invalid compaction")]
    InvalidCompaction,

    #[error("compaction executor failed")]
    CompactionExecutorFailed,

    #[error(
        "invalid clock tick, must be monotonic. last_tick=`{last_tick}`, next_tick=`{next_tick}`"
    )]
    InvalidClockTick { last_tick: i64, next_tick: i64 },

    #[error("detected newer DB client")]
    Fenced,

    #[error("invalid cache part size bytes, it must be multiple of 1024 and greater than 0")]
    InvalidCachePartSize,

    #[error("invalid compression codec")]
    InvalidCompressionCodec,

    #[cfg(any(
        feature = "snappy",
        feature = "zlib",
        feature = "lz4",
        feature = "zstd"
    ))]
    #[error("error decompressing block")]
    BlockDecompressionError,

    #[cfg(any(feature = "snappy", feature = "zlib", feature = "zstd"))]
    #[error("error compressing block")]
    BlockCompressionError,

    #[error("Invalid RowFlags. #{message}. encoded_bits=`{encoded_bits:#b}`, known_bits=`{known_bits:#b}`"
    )]
    InvalidRowFlags {
        encoded_bits: u8,
        known_bits: u8,
        message: String,
    },

    #[error("read channel error")]
    ReadChannelError(#[from] tokio::sync::oneshot::error::RecvError),

    #[error("iterator invalidated after unexpected error")]
    InvalidatedIterator(#[from] Box<SlateDBError>),

    #[error("background task panic'd")]
    // we need to wrap the panic args in an Arc so SlateDbError is Clone
    // we need to wrap the panic args in a mutex so that SlateDbError is Sync
    BackgroundTaskPanic(Arc<Mutex<Box<dyn Any + Send>>>),

    #[error("background task shutdown")]
    BackgroundTaskShutdown,

    #[error("merge operator error")]
    MergeOperatorError(#[from] MergeOperatorError),

    #[error("checkpoint missing. checkpoint_id=`{0}`")]
    CheckpointMissing(Uuid),

    #[error(
        "byte format version mismatch. expected_version=`{expected_version}`, actual_version=`{actual_version}`"
    )]
    InvalidVersion {
        expected_version: u16,
        actual_version: u16,
    },

    #[error("foyer cache reading error")]
    #[cfg(feature = "foyer")]
    FoyerCacheReadingError(#[from] Arc<anyhow::Error>),

    #[error("operation timed out. operation=`{op}`")]
    Timeout { op: &'static str, backoff: Duration },

    #[error("wal buffer already started")]
    WalBufferAlreadyStarted,

    #[error("cannot seek to a key outside the iterator range. key=`{key:?}`, range=`{range:?}`")]
    SeekKeyOutOfRange { key: Vec<u8>, range: BytesRange },

    #[error("cannot seek to a key less than the last returned key")]
    SeekKeyLessThanLastReturnedKey,

    #[error(
        "parent path must be different from the clone's path. parent_path=`{0}`, clone_path=`{0}`"
    )]
    IdenticalClonePaths(Path),

    #[error("invalid checkpoint lifetime. lifetime=`{0:?}`")]
    InvalidCheckpointLifetime(Duration),

    #[error("invalid manifest poll interval. interval=`{0:?}`")]
    InvalidManifestPollInterval(Duration),

    #[error("checkpoint lifetime must be at least double the manifest poll interval. lifetime=`{lifetime:?}`, interval=`{interval:?}`")]
    CheckpointLifetimeTooShort {
        lifetime: Duration,
        interval: Duration,
    },

    #[error("invalid sst batch size. size=`{0}`")]
    InvalidSSTBatchSize(usize),

    #[error("cannot seek to a key outside the iterator range. key=`{key:?}`, start_key=`{start_key:?}`, end_key=`{end_key:?}`")]
    SeekKeyOutOfKeyRange {
        key: Vec<u8>,
        start_key: Bound<Vec<u8>>,
        end_key: Bound<Vec<u8>>,
    },

    #[error("the cloned database is not attached to any external database")]
    CloneExternalDbMissing,

    #[error("the cloned database is not attached to external database with a valid checkpoint. path=`{path}`, checkpoint_id=`{checkpoint_id:?}`")]
    CloneIncorrectExternalDbCheckpoint {
        path: String,
        checkpoint_id: Option<Uuid>,
    },

    #[error("the final checkpoint for the cloned database no longer exists in the manifest. path=`{path}`, checkpoint_id=`{checkpoint_id}`")]
    CloneIncorrectFinalCheckpoint { path: String, checkpoint_id: Uuid },

    #[error("unknown configuration file format. path=`{0}`")]
    UnknownConfigurationFormat(PathBuf),

    #[error("invalid configuration format")]
    InvalidConfigurationFormat(#[from] Box<figment::Error>),

    #[error("attempted a WAL operation when the WAL is disabled")]
    WalDisabled,

    #[error("invalid object store URL. url=`{0}`")]
    InvalidObjectStoreURL(String, #[source] url::ParseError),
}

impl From<std::io::Error> for SlateDBError {
    fn from(value: std::io::Error) -> Self {
        Self::IoError(Arc::new(value))
    }
}

impl From<object_store::Error> for SlateDBError {
    fn from(value: object_store::Error) -> Self {
        Self::ObjectStoreError(Arc::new(value))
    }
}

type BoxError = Box<dyn std::error::Error + Send + Sync>;

/// Represents the kind of public errors that can be returned to the user.
#[derive(Debug)]
pub enum ErrorKind {
    /// The database attempted to use an invalid configuration.
    Configuration,

    /// The database attempted an invalid operation or an operation with an
    /// invalid parameter (including misconfiguration).
    Operation,

    /// Unexpected internal error. This error is fatal (i.e. the database must be closed).
    System,

    /// Invalid persistent state (e.g. corrupted data files). The state must
    /// be repaired before the database can be restarted.
    PersistentState,

    /// Failed access database resources (e.g. remote storage) due to some
    /// kind of authentication or authorization error. The operation can be
    /// retried after the permission issue is resolved.
    Permission,

    /// The operation failed due to a transient error (such as IO unavailability).
    /// The operation can be retried after backing off.
    Transient(Duration),
}

impl std::fmt::Display for ErrorKind {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            ErrorKind::Configuration => write!(f, "Configuration error"),
            ErrorKind::Operation => write!(f, "Operation error"),
            ErrorKind::System => write!(f, "System error"),
            ErrorKind::PersistentState => write!(f, "Persistent state error"),
            ErrorKind::Permission => write!(f, "Permission error"),
            ErrorKind::Transient(backoff) => {
                write!(f, "Transient error (retry after: {backoff:?})")
            }
        }
    }
}

/// Represents a public error that can be returned to the user.
#[derive(Debug)]
pub struct Error {
    msg: String,
    kind: ErrorKind,
    source: Option<BoxError>,
}

impl std::fmt::Display for Error {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}: {}", self.kind, self.msg)?;
        if let Some(source) = self.source.as_ref() {
            write!(f, " ({source})")?;
        }
        Ok(())
    }
}

impl std::error::Error for Error {
    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
        self.source
            .as_ref()
            .map(|e| e.as_ref() as &(dyn std::error::Error + 'static))
    }
}

impl Error {
    /// Creates a new public configuration error.
    pub fn configuration(msg: String) -> Self {
        Self {
            msg,
            kind: ErrorKind::Configuration,
            source: None,
        }
    }

    /// Creates a new public operation error.
    pub fn operation(msg: String) -> Self {
        Self {
            msg,
            kind: ErrorKind::Operation,
            source: None,
        }
    }

    /// Creates a new public system error.
    pub fn system(msg: String) -> Self {
        Self {
            msg,
            kind: ErrorKind::System,
            source: None,
        }
    }

    /// Creates a new public persistent state error.
    pub fn persistent_state(msg: String) -> Self {
        Self {
            msg,
            kind: ErrorKind::PersistentState,
            source: None,
        }
    }

    /// Creates a new public permission error.
    pub fn permission(msg: String) -> Self {
        Self {
            msg,
            kind: ErrorKind::Permission,
            source: None,
        }
    }

    /// Creates a new public transient error.
    pub fn transient(msg: String, backoff: Duration) -> Self {
        Self {
            msg,
            kind: ErrorKind::Transient(backoff),
            source: None,
        }
    }

    /// Adds a source to the error.
    pub fn with_source(mut self, source: BoxError) -> Self {
        self.source = Some(source);
        self
    }
}

impl From<SlateDBError> for Error {
    fn from(err: SlateDBError) -> Self {
        let msg = err.to_string();
        match err {
            SlateDBError::IoError(err) => Error::system(msg).with_source(Box::new(err)),
            SlateDBError::ObjectStoreError(err) => Error::operation(msg).with_source(Box::new(err)),
            SlateDBError::InvalidFlatbuffer(err) => {
                Error::persistent_state(msg).with_source(Box::new(err))
            }
            SlateDBError::InvalidDeletion => Error::operation(msg),
            SlateDBError::InvalidDBState => Error::persistent_state(msg),
            SlateDBError::InvalidCompaction => Error::operation(msg),
            SlateDBError::CompactionExecutorFailed => Error::system(msg),
            SlateDBError::InvalidClockTick { .. } => Error::operation(msg),
            SlateDBError::Fenced => Error::permission(msg),
            SlateDBError::InvalidCachePartSize => Error::operation(msg),
            SlateDBError::InvalidCompressionCodec => Error::operation(msg),
            #[cfg(any(
                feature = "snappy",
                feature = "zlib",
                feature = "lz4",
                feature = "zstd"
            ))]
            SlateDBError::BlockDecompressionError => Error::operation(msg),
            #[cfg(any(feature = "snappy", feature = "zlib", feature = "zstd"))]
            SlateDBError::BlockCompressionError => Error::operation(msg),
            SlateDBError::InvalidRowFlags { .. } => Error::operation(msg),
            SlateDBError::ReadChannelError(err) => Error::system(msg).with_source(Box::new(err)),
            SlateDBError::InvalidatedIterator(err) => {
                Error::operation(msg).with_source(Box::new(err))
            }
            SlateDBError::BackgroundTaskPanic(err) => {
                Error::system(msg).with_source(Box::new(PanicError(err)))
            }
            SlateDBError::BackgroundTaskShutdown => Error::system(msg),
            SlateDBError::MergeOperatorError(err) => {
                Error::operation(msg).with_source(Box::new(err))
            }
            SlateDBError::CheckpointMissing(_) => Error::persistent_state(msg),
            SlateDBError::InvalidVersion { .. } => Error::persistent_state(msg),
            #[cfg(feature = "foyer")]
            SlateDBError::FoyerCacheReadingError(err) => {
                Error::persistent_state(msg).with_source(Box::new(AnyhowError(err)))
            }
            SlateDBError::Timeout { backoff, .. } => Error::transient(msg, backoff),
            SlateDBError::WalBufferAlreadyStarted => Error::system(msg),
            SlateDBError::ManifestMissing(_) => Error::persistent_state(msg),
            SlateDBError::LatestManifestMissing => Error::persistent_state(msg),
            SlateDBError::ManifestVersionExists => Error::persistent_state(msg),
            SlateDBError::EmptyManifest => Error::persistent_state(msg),
            SlateDBError::EmptyBlock => Error::persistent_state(msg),
            SlateDBError::EmptyBlockMeta => Error::persistent_state(msg),
            SlateDBError::EmptySSTable => Error::persistent_state(msg),
            SlateDBError::Unsupported(_) => Error::operation(msg),
            SlateDBError::ChecksumMismatch => Error::persistent_state(msg),
            SlateDBError::SeekKeyOutOfRange { .. } => Error::operation(msg),
            SlateDBError::SeekKeyLessThanLastReturnedKey => Error::operation(msg),
            SlateDBError::IdenticalClonePaths { .. } => Error::operation(msg),
            SlateDBError::InvalidCheckpointLifetime(_) => Error::operation(msg),
            SlateDBError::InvalidManifestPollInterval(_) => Error::operation(msg),
            SlateDBError::CheckpointLifetimeTooShort { .. } => Error::operation(msg),
            SlateDBError::InvalidSSTBatchSize(_) => Error::operation(msg),
            SlateDBError::SeekKeyOutOfKeyRange { .. } => Error::operation(msg),
            SlateDBError::CloneExternalDbMissing => Error::persistent_state(msg),
            SlateDBError::CloneIncorrectExternalDbCheckpoint { .. } => Error::persistent_state(msg),
            SlateDBError::CloneIncorrectFinalCheckpoint { .. } => Error::persistent_state(msg),
            SlateDBError::UnknownConfigurationFormat(_) => Error::configuration(msg),
            SlateDBError::InvalidConfigurationFormat(err) => {
                Error::configuration(msg).with_source(Box::new(err))
            }
            SlateDBError::WalDisabled => Error::operation(msg),
            SlateDBError::InvalidObjectStoreURL(_, err) => {
                Error::configuration(msg).with_source(Box::new(err))
            }
        }
    }
}

#[derive(Debug)]
struct PanicError(Arc<Mutex<Box<dyn Any + Send>>>);
impl std::error::Error for PanicError {}
impl std::fmt::Display for PanicError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let guard = self.0.lock().expect("Failed to lock panic error");
        if let Some(err) = guard.downcast_ref::<SlateDBError>() {
            write!(f, "{err}")?;
        } else if let Some(err) = guard.downcast_ref::<Box<dyn std::error::Error>>() {
            write!(f, "{err}")?;
        } else if let Some(err) = guard.downcast_ref::<String>() {
            write!(f, "{err}")?;
        } else {
            write!(f, "irrecoverable panic")?;
        }
        Ok(())
    }
}

#[derive(Debug)]
#[cfg(feature = "foyer")]
struct AnyhowError(Arc<anyhow::Error>);
#[cfg(feature = "foyer")]
impl std::error::Error for AnyhowError {}
#[cfg(feature = "foyer")]
impl std::fmt::Display for AnyhowError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.0)
    }
}
