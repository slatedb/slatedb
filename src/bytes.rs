use bytes::{BufMut, Bytes, BytesMut};

#[derive(Clone)]
pub(crate) struct OrderedBytesGenerator {
    suffix: Bytes,
    bytes: Vec<u8>,
    min: u8,
    max: u8,
}

impl OrderedBytesGenerator {
    #[cfg(feature = "bencher")]
    pub(crate) fn new_with_suffix(suffix: &[u8], bytes: &[u8]) -> Self {
        Self::new(suffix, bytes, u8::MIN, u8::MAX)
    }

    #[cfg(test)]
    pub(crate) fn new_with_byte_range(bytes: &[u8], min: u8, max: u8) -> Self {
        Self::new(&[], bytes, min, max)
    }

    pub(crate) fn new(suffix: &[u8], bytes: &[u8], min: u8, max: u8) -> Self {
        let bytes = Vec::from(bytes);
        Self {
            suffix: Bytes::copy_from_slice(suffix),
            bytes,
            min,
            max,
        }
    }

    pub(crate) fn next(&mut self) -> Bytes {
        let mut result = BytesMut::with_capacity(self.bytes.len() + std::mem::size_of::<u32>());
        result.put_slice(self.bytes.as_slice());
        result.put(self.suffix.as_ref());
        self.increment();
        result.freeze()
    }

    fn increment(&mut self) {
        let mut pos = self.bytes.len() - 1;
        while self.bytes[pos] == self.max {
            self.bytes[pos] = self.min;
            pos -= 1;
        }
        self.bytes[pos] += 1;
    }
}

#[cfg(test)]
mod tests {
    use bytes::{BufMut, Bytes};

    use crate::bytes::OrderedBytesGenerator;

    #[test]
    fn test_should_generate_ordered_bytes() {
        let mut suffix = Vec::<u8>::new();
        suffix.put_u32(3735928559);
        let start = [0u8, 0u8, 0u8];
        let mut gen = OrderedBytesGenerator::new(suffix.as_ref(), &start, 0, 2);

        let expected = [
            [0u8, 0u8, 0u8, 0xde, 0xad, 0xbe, 0xef],
            [0u8, 0u8, 1u8, 0xde, 0xad, 0xbe, 0xef],
            [0u8, 0u8, 2u8, 0xde, 0xad, 0xbe, 0xef],
            [0u8, 1u8, 0u8, 0xde, 0xad, 0xbe, 0xef],
            [0u8, 1u8, 1u8, 0xde, 0xad, 0xbe, 0xef],
            [0u8, 1u8, 2u8, 0xde, 0xad, 0xbe, 0xef],
            [0u8, 2u8, 0u8, 0xde, 0xad, 0xbe, 0xef],
            [0u8, 2u8, 1u8, 0xde, 0xad, 0xbe, 0xef],
            [0u8, 2u8, 2u8, 0xde, 0xad, 0xbe, 0xef],
            [1u8, 0u8, 0u8, 0xde, 0xad, 0xbe, 0xef],
            [1u8, 0u8, 1u8, 0xde, 0xad, 0xbe, 0xef],
            [1u8, 0u8, 2u8, 0xde, 0xad, 0xbe, 0xef],
            [1u8, 1u8, 0u8, 0xde, 0xad, 0xbe, 0xef],
            [1u8, 1u8, 1u8, 0xde, 0xad, 0xbe, 0xef],
            [1u8, 1u8, 2u8, 0xde, 0xad, 0xbe, 0xef],
            [1u8, 2u8, 0u8, 0xde, 0xad, 0xbe, 0xef],
            [1u8, 2u8, 1u8, 0xde, 0xad, 0xbe, 0xef],
            [1u8, 2u8, 2u8, 0xde, 0xad, 0xbe, 0xef],
            [2u8, 0u8, 0u8, 0xde, 0xad, 0xbe, 0xef],
            [2u8, 0u8, 1u8, 0xde, 0xad, 0xbe, 0xef],
            [2u8, 0u8, 2u8, 0xde, 0xad, 0xbe, 0xef],
            [2u8, 1u8, 0u8, 0xde, 0xad, 0xbe, 0xef],
            [2u8, 1u8, 1u8, 0xde, 0xad, 0xbe, 0xef],
            [2u8, 1u8, 2u8, 0xde, 0xad, 0xbe, 0xef],
            [2u8, 2u8, 0u8, 0xde, 0xad, 0xbe, 0xef],
            [2u8, 2u8, 1u8, 0xde, 0xad, 0xbe, 0xef],
        ];
        for e in expected.iter() {
            assert_eq!(gen.next(), Bytes::copy_from_slice(e))
        }
    }
}
