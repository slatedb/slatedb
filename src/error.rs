#[derive(thiserror::Error, Debug)]
pub enum SlateDBError {
    #[error("IO error")]
    IoError(#[from] std::io::Error),

    #[error("Checksum mismatch")]
    ChecksumMismatch,

    #[error("Empty SSTable")]
    EmptySSTable,

    #[error("Empty block metadata")]
    EmptyBlockMeta,

    #[error("Empty block")]
    EmptyBlock,

    #[error("Object store error")]
    ObjectStoreError(#[from] object_store::Error),

    #[error("Manifest file already exists")]
    ManifestVersionExists,

    #[error("Manifest missing")]
    ManifestMissing,

    #[error("Invalid deletion")]
    InvalidDeletion,

    #[error("Invalid sst error")]
    InvalidFlatbuffer(#[from] flatbuffers::InvalidFlatbuffer),

    #[error("Invalid DB state error")]
    InvalidDBState,

    #[error("Invalid Compaction")]
    InvalidCompaction,

    #[error("Detected newer DB client")]
    Fenced,

    #[error("Invalid cache part size bytes, it must be multiple of 1024 and greater than 0")]
    InvalidCachePartSize,

    #[error("Invalid Compression Codec")]
    InvalidCompressionCodec,

    #[error("Error Decompressing Block")]
    BlockDecompressionError,

    #[error("Error Compressing Block")]
    BlockCompressionError,
}
