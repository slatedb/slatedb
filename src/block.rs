use bytes::{Buf, BufMut, Bytes, BytesMut};

use crate::error::SlateDBError;

pub(crate) const SIZEOF_U16: usize = std::mem::size_of::<u16>();
pub(crate) const SIZEOF_U32: usize = std::mem::size_of::<u32>();

/// "None" values are encoded by using the maximum u32 value as the value length.
pub(crate) const TOMBSTONE: u32 = u32::MAX;

pub struct Block {
    pub(crate) data: Bytes,
    pub(crate) offsets: Vec<u16>,
}

impl Block {
    #[rustfmt::skip]
    pub fn encode(&self) -> Bytes {
        let mut buf = BytesMut::with_capacity(
            self.data.len()                   // data byte length
            + self.offsets.len() * SIZEOF_U16 // offsets as u16's
            + SIZEOF_U16, // number of offsets in the block
        );
        buf.put_slice(&self.data);
        for offset in &self.offsets {
            buf.put_u16(*offset);
        }
        buf.put_u16(self.offsets.len() as u16);
        buf.freeze()
    }

    #[rustfmt::skip]
    pub fn decode(bytes: Bytes) -> Self {
        // Get number of elements in the block
        let data = bytes.as_ref();
        let entry_offsets_len = (&data[data.len() - SIZEOF_U16..]).get_u16() as usize;
        let data_end = data.len()
            - SIZEOF_U16                                            // Entry u16 length
            - entry_offsets_len * SIZEOF_U16; // Offset byte array length
        let offsets_raw = &data[data_end..data.len() - SIZEOF_U16]; // Entry u16
        let offsets = offsets_raw
            .chunks(SIZEOF_U16)
            .map(|mut x| x.get_u16())
            .collect();
        let bytes = bytes.slice(0..data_end);
        Self {
            data: bytes,
            offsets,
        }
    }
}

pub struct BlockBuilder {
    offsets: Vec<u16>,
    data: Vec<u8>,
    block_size: usize,
}

impl BlockBuilder {
    pub fn new(block_size: usize) -> Self {
        Self {
            offsets: Vec::new(),
            data: Vec::new(),
            block_size,
        }
    }

    #[rustfmt::skip]
    fn estimated_size(&self) -> usize {
        SIZEOF_U16           // number of key-value pairs in the block
        + self.offsets.len() // offsets
        + self.data.len()    // key-value pairs
    }

    #[must_use]
    pub fn add(&mut self, key: &[u8], value: Option<&[u8]>) -> bool {
        assert!(!key.is_empty(), "key must not be empty");

        // If adding the key-value pair would exceed the block size limit, don't add it.
        // (Unless the block is empty, in which case, allow the block to exceed the limit.)
        if self.estimated_size()
                + key.len()
                + value.map(|v| v.len()).unwrap_or_default() // None takes no space (besides u32)
                + SIZEOF_U16 * 2 // key size and offset size
                + SIZEOF_U32 // value size
                > self.block_size
            && !self.is_empty()
        {
            return false;
        }

        self.offsets.push(self.data.len() as u16);
        self.data.put_u16(key.len() as u16);
        self.data.put(key);
        if let Some(value) = value {
            self.data.put_u32(value.len() as u32);
            self.data.put(value);
        } else {
            self.data.put_u32(TOMBSTONE);
        }

        true
    }

    pub fn is_empty(&self) -> bool {
        self.offsets.is_empty()
    }

    pub fn build(self) -> Result<Block, SlateDBError> {
        if self.is_empty() {
            return Err(SlateDBError::EmptyBlock);
        }
        Ok(Block {
            data: Bytes::from(self.data),
            offsets: self.offsets,
        })
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_block() {
        let mut builder = BlockBuilder::new(4096);
        assert!(builder.add(b"key1", Some(b"value1")));
        assert!(builder.add(b"key2", Some(b"value2")));
        let block = builder.build().unwrap();
        let encoded = block.encode();
        let decoded = Block::decode(encoded);
        assert_eq!(block.data, decoded.data);
        assert_eq!(block.offsets, decoded.offsets);
    }

    #[test]
    fn test_block_with_tombstone() {
        let mut builder = BlockBuilder::new(4096);
        assert!(builder.add(b"key1", Some(b"value1")));
        assert!(builder.add(b"key2", None));
        assert!(builder.add(b"key3", Some(b"value3")));
        let block = builder.build().unwrap();
        let encoded = block.encode();
        let _decoded = Block::decode(encoded);
    }
}
