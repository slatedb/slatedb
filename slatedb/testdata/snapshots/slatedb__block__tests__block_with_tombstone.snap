---
source: src/block.rs
expression: "(block.size(), block.data, block.offsets)"
snapshot_kind: text
---
(
    113,
    b"\0\0\0\x04key1\0\0\0\0\0\0\0\0\x06\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\x06value1\0\x03\0\x012\0\0\0\0\0\0\0\0\x05\0\0\0\0\0\0\0\0\0\x03\0\x013\0\0\0\0\0\0\0\0\x06\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\x06value3",
    [
        0,
        43,
        65,
    ],
)
